framework/include/interfaces/TaggingInterface.h

```diff
@@ -44,18 +51,54 @@ class TaggingInterface
 
   bool isMatrixTagged() { return _matrix_tags.size() > 0; }
 
-  const std::set<TagID> & getVectorTags() const { return _vector_tags; }
+  /**
+   * Class that is used as a parameter to getVectorTags() that allows only
+   * AttribVectorTags methods to call it
+   */
+  class GetVectorTagsKey
+  {
+    friend class AttribVectorTags;
+    friend class NonlinearEigenSystem;
+    template <typename>
+    friend class MooseObjectTagWarehouse;
+
+    GetVectorTagsKey() {}
+    GetVectorTagsKey(const GetVectorTagsKey &) {}
+  };
 
-  const std::set<TagID> & getMatrixTags() const { return _matrix_tags; }
+  const std::set<TagID> & getVectorTags(GetVectorTagsKey) const { return _vector_tags; }
 
   /**
-   * Prepare data for computing element residual the according to active tags.
+   * Class that is used as a parameter to getMatrixTags() that allows only
+   * AttribMatrixTags methods to call it
+   */
+  class GetMatrixTagsKey
+  {
+    friend class AttribMatrixTags;
+    friend class NonlinearEigenSystem;
+    template <typename>
+    friend class MooseObjectTagWarehouse;
+
+    GetMatrixTagsKey() {}
+    GetMatrixTagsKey(const GetMatrixTagsKey &) {}
+  };
+
+  const std::set<TagID> & getMatrixTags(GetMatrixTagsKey) const { return _matrix_tags; }
+
+protected:
+  /**
+   * Prepare data for computing element residual according to active tags.
    * Residual blocks for different tags will be extracted from Assembly.
    * A local residual will be zeroed. It should be called
    * right before the local element vector is computed.
    */
   void prepareVectorTag(Assembly & assembly, unsigned int ivar);
 
+  void prepareVectorTag(Assembly & assembly,
+                        unsigned int ivar,
+                        const ReferenceResidualProblem * ref_problem,
+                        bool prepare_non_ref_tags);
```

Comment:
Add doxygen. Also maybe either
1. change to `bool prepare_ref_tags` (no negation)
2. Or make this an enum